src/core/execute.c

```diff
@@ -3223,6 +3223,15 @@ static int exec_child(
         else
                 needs_setuid = (params->flags & EXEC_APPLY_SANDBOXING) && !(command->flags & (EXEC_COMMAND_FULLY_PRIVILEGED|EXEC_COMMAND_NO_SETUID));
 
+        /* Drop groups as early as possbile */
+        if (needs_setuid) {
+                r = enforce_groups(gid, supplementary_gids, ngids);
+                if (r < 0) {
+                        *exit_status = EXIT_GROUP;
+                        return log_unit_error_errno(unit, r, "Changing group credentials failed: %m");
+                }
+        }
```

Comment:
hmm, this moves gid initialization before PAM. That's quite a change, and I am not sure this is how PAM is supposed to be used. IIRC the rule for PAM session hooks is that they are invoked before setresuid()/setresgid()... This would change this.

Reply:
I guess you mean it's done before setup_pam? I don't see any reason why this cannot be after setup_pam. It just needs to be done before the chroot.

Reply:
PAM generally assumes its modules are run with full privs, hence doing setresuid()/setresguid()/ beforehand appears problematic to me.

Reply:
PAM needs to be done before the chroot. enforce_groups too. PAM needs to be done before enforce_groups. So the correct order is setup_pam -> enforce_groups -> apply_mount_namespace. As far as I understand, this would fully address your concern. If not, I have misunderstood it.

Reply:
yeah, that sounds ok.

another option btw, is to split maybe_setgroups() so that the result of the /proc check is done early, and the setgroups() call later. that might be nicer no, as it would make the changes to execute.c more minimal?